src/core/backend_context.cc

```diff
@@ -165,34 +165,26 @@ BackendContext::SetFixedSizeOutputBuffer(
     if (payload.status_.IsOk() && (payload.response_provider_ != nullptr) &&
         payload.response_provider_->RequiresOutput(name)) {
       auto dst_memory_type = src_memory_type;
+      int64_t dst_device_id;
       void* buffer = nullptr;
 
       // try to get buffer with the same memory type as the output tensor
       Status status = payload.response_provider_->AllocateOutputBuffer(
           name, &buffer, expected_byte_size, content_shape, src_memory_type,
-          src_memory_type_id);
-
+          src_memory_type_id, &dst_memory_type, &dst_device_id);
+      LOG_VERBOSE(1) << "src_memory_type: " << src_memory_type;
+      LOG_VERBOSE(1) << "dst_memory_type: " << dst_memory_type;
       if (status.IsOk() && (expected_byte_size != 0)) {
-        if ((buffer == nullptr) && (src_memory_type != TRTSERVER_MEMORY_CPU)) {
-          // Use default (CPU memory type) if preferred type can't be fulfilled
-          status = payload.response_provider_->AllocateOutputBuffer(
-              name, &buffer, expected_byte_size, content_shape);
-          dst_memory_type = TRTSERVER_MEMORY_CPU;
-        }
-
-        if (status.IsOk()) {
-          if (buffer == nullptr) {
-            status = Status(
-                RequestStatusCode::INTERNAL,
-                "all attempts to allocate buffer for output '" + name +
-                    "' failed");
-          } else {
-            bool cuda_used = false;
-            status = CopyBuffer(
-                name, src_memory_type, dst_memory_type, expected_byte_size,
-                content + content_offset, buffer, &cuda_used);
-            cuda_copy |= cuda_used;
-          }
+        if (buffer == nullptr) {
+          status = Status(
+              RequestStatusCode::INTERNAL,
+              "failed to allocate buffer for output '" + name + "'");
+        } else {
+          bool cuda_used = false;
+          status = CopyBuffer(
+              name, src_memory_type, dst_memory_type, expected_byte_size,
```

Comment:
How can we copy correctly without knowing the memory_type_ids?

Reply:
cudaMemcpyDeviceToDevice is used. According to the docs it is able to detect source and destination GPU device IDS based on the address. @GuanLuo can you confirm?

Reply:
It appears that DeviceToDevice copy should only be used for copies on the same GPU. For different GPU devices, `cudaMemcpyPeerAsync` should be used. 
Here we should also check if PeerAccess is allowed between those two GPUs else do DeviceToHost + HostToDevice (don't see a better route) I believe this would tie into the tensor copy logic cleanup.

Reply:
I didn't realize `cudaMemcpyDeviceToDevice` flag is limited to copy within the same device, this may explain the bug that I am working on. We should address this ASAP as data transfer between GPUs are clearly broken.

Doing  "DeviceToHost + HostToDevice" should be the default choice. But it seems to have too much an overhead. Then we can optimize in the case if PeerAccess is allowed. And in the case that it's not, we should warn user that using CUDA shared memory across devices is not suggested and for ensemble, the intermediate data should always be allocated in system memory (until it gets smarter).

Reply:
This makes sense. What I have done right now should be enough for now. We can add Peer Access to avoid transfer in the copy tensor cleanup ticket